test/Test.hs

```diff
@@ -0,0 +1,132 @@
+      ]
+  , testGroup "strict handler with raw MultipartData"
+      [ testWai testApp "correct body" testBlogPostRawHandler
+      ]
+  ]
+
+data BlogPost
+  = BlogPost
+      { title :: Text
+      , body  :: Text
+      }
+
+instance FromMultipart Mem BlogPost where
+  fromMultipart md =
+    BlogPost
+      <$> lookupInput "title" md
+      <*> fmap (cs . fdPayload) (lookupFile "body" md)
+
+type TestAPI
+  =    "blogPostStrict" :> MultipartForm Mem BlogPost :> Post '[PlainText] Text
+  :<|> "blogPostLenient" :> MultipartForm' '[Lenient] Mem BlogPost :> Post '[JSON] Bool
+  :<|> "blogPostRaw" :> MultipartForm Mem (MultipartData Mem) :> Post '[PlainText] Text
+
+blogPostStrictHandler :: BlogPost -> Handler Text
+blogPostStrictHandler bp = return $ title bp <> "\n" <> body bp
+
+blogPostLenientHandler :: Either String BlogPost -> Handler Bool
+blogPostLenientHandler eitherBP =
+  case eitherBP of
+    Left _  -> return False
+    Right _ -> return True
+
+blogPostRawHandler :: MultipartData Mem -> Handler Text
+blogPostRawHandler md =
+  return $ mconcat $ intersperse " "
+    $ map iName (inputs md) <> map fdInputName (files md)
+
+testApp :: Application
+testApp = serve @TestAPI Proxy $ blogPostStrictHandler :<|> blogPostLenientHandler :<|> blogPostRawHandler
+
+multipartHeaders :: [(HeaderName, BS.ByteString)]
+multipartHeaders = [(hContentType, "multipart/form-data; boundary=XX")]
+
+testBlogPostStrictHandler :: Session ()
+testBlogPostStrictHandler = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostStrict" correctBody multipartHeaders
+  assertStatus 200 res
+  assertBody "Foo post\nFoo body\n" res
+
+testBlogPostStrictHandlerEmptyBody :: Session ()
+testBlogPostStrictHandlerEmptyBody = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostStrict" "" multipartHeaders
+  assertStatus 400 res
+  assertBody "fromMultipart failed: Field title not found" res
+
+testBlogPostStrictHandlerPartialBody :: Session ()
+testBlogPostStrictHandlerPartialBody = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostStrict" partialBody multipartHeaders
+  assertStatus 400 res
+  assertBody "fromMultipart failed: File body not found" res
+
+testBlogPostLenientHandler :: Session ()
+testBlogPostLenientHandler = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostLenient" correctBody multipartHeaders
+  assertStatus 200 res
+  assertBody "true" res
+
+testBlogPostLenientHandlerPartialBody :: Session ()
+testBlogPostLenientHandlerPartialBody = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostLenient" partialBody multipartHeaders
+  assertStatus 200 res
+  assertBody "false" res
+
+testBlogPostRawHandler :: Session ()
+testBlogPostRawHandler = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostRaw" correctBody multipartHeaders
+  assertStatus 200 res
+  assertBody "title body" res
+
+correctBody :: BSL.ByteString
+correctBody = mconcat $ intersperse "\n"
```

Comment:
again, feel free to ignore: marginally nicer (or worse, depending on your taste :) -

```suggestion
correctBody = cs $ unlines
```